### PR TITLE
feat(python): add entry point to the Consortium DataFrame API

### PR DESCRIPTION
**Pull request by MarcoGorelli**
Here's an entry point to the Consortium's DataFrame API Standard

It enables dataframe-consuming libraries to just check for a `__dataframe_consortium_standard__` attribute on a dataframe they receive - then, so long as they stick to the spec defined in https://data-apis.org/dataframe-api/draft/index.html, then their code should work the same way, regardless of what the original backing dataframe library was

Use-case: currently, scikit-learn is very keen on using this, as they're not keen on having to depend on pyarrow (which will become required in pandas), and the interchange protocol only goes so far (e.g. a way to convert to ndarray is out-of-scope for that). If we can get this to work there, then other use cases may emerge

**Maintenance burden on polars: none** (unless, once it gains traction, you want it upstreamed). The compat package will be developed, maintained, and tested by the consortium and community of libraries which use it. It is up to consuming libraries to set a minimum version of the dataframe-api-compat package

The current spec should be enough for scikit-learn, and having this entry point makes it easier to move forwards with development (without monkey-patching / special-casing)

### CODE REVIEW

File: py-polars/polars/dataframe/frame.py
Comment (stinodego) [inline py-polars/polars/dataframe/frame.py:1239]: I don't think we need the `/` indicator here?

```suggestion
        self, *, api_version: str | None = None
```

File: py-polars/polars/lazyframe/frame.py
Comment (stinodego) [inline py-polars/polars/lazyframe/frame.py:643]: Does this function check for `LazyFrame`? Or how does this work exactly?

If it collects behind the scenes, I think we should not allow this method on LazyFrames and only have it on DataFrames.

File: py-polars/polars/dataframe/frame.py
Comment (stinodego) [inline py-polars/polars/dataframe/frame.py:1253]: Should this be part of our `dependencies` module?

**Review by stinodego — COMMENTED** [review]
This is very interesting. Does that mean we do not implement the interchange protocol within the library anymore, but implement it in the external package?

I was planning to work on this the coming days, so your timing is impeccable 🎉

**Comment by MarcoGorelli** [comment]
> so your timing is impeccable 🎉

😄 sorry for the confusion, this is separate (but related!) to the interchange protocol

The interchange protocol is intended more for converting between dataframe libraries. This is OK for plotting libraries that don't do any heavy lifting (e.g. that's what plotly does)

The standard is more an attempt to agree on a minimal set of "essential dataframe functionality", which would work the same way across dataframe libraries (in separate namespaces, as at this point there's no chance of getting everyone to agree on having the same main namespace), which would allow you to write something like
```python
def remove_outliers(df, column_name):
    # Get a Standard-compliant dataframe.
    # NOTE: this has not yet been upstreamed, so won't work out-of-the-box!
    # See 'resources' below for how to try it out.
    df_standard = df.__dataframe_consortium_standard__()
    # Use methods from the Standard specification.
    col = df_standard.get_column_by_name(column_name)
    z_score = (col - col.mean()) / col.std()
    df_standard_filtered = df_standard.get_rows_by_mask((z_score > -3) & (z_score < 3))
    # Return the result as a dataframe from the original library.
    return df_standard_filtered.dataframe
```
and have it work for any `df` which has a namespace implementing the Consortium's standard (without any conversions between libraries, nor heavy dependencies like pyarrow)

So, it doesn't replace the interchange protocol, but rather it enables things which are out-of-scope for the interchange protocol

I plan to make a similar PR to pandas after this. Other libraries will follow, but these will already be enough to get things moving in scikit-learn

File: py-polars/polars/lazyframe/frame.py
Comment (MarcoGorelli) [reply py-polars/polars/lazyframe/frame.py:643]: yup, the aim is that it _should_ all work for lazyframes - if you start with a lazyframe, the operation will return a lazyframe

In practice, the implementation isn't there yet, but that can be handled there (and libraries can then just set a minimum version on `dataframe-api-compat`, which should be easier than waiting for everything to be 100% and having to set a very high minimum version on polars itself)

File: py-polars/polars/dependencies.py
Comment (stinodego) [inline py-polars/polars/dependencies.py:226]: I think this should then also become part of our development dependencies / extras defined in pyproject.toml, and maybe include a very simple test for each just to show that the methods work / don't get deleted accidentally?

**Comment by stinodego** [comment]
Thanks for the explanation; makes sense!

File: py-polars/polars/dependencies.py
Comment (MarcoGorelli) [reply py-polars/polars/dependencies.py:226]: good idea, thanks - I've added to `requirements-dev.txt`. Leaving it out of `pyproject.toml` for now, perhaps it can be added there once it's more mature?

Comment (stinodego) [reply py-polars/polars/dependencies.py:226]: Fair enough!

File: py-polars/tests/unit/test_consortium_standard.py
Comment (stinodego) [inline py-polars/tests/unit/test_consortium_standard.py:11]: We shouldn't be skipping tests like this. The functionality is in the library, so it must be tested. You'll just have to make sure your dev environment is up to date.

**Review by stinodego — COMMENTED** [review]
Just one final comment, then it has my blessing!

**Review by stinodego — APPROVED** [review]
Good to go if the CI passes!

File: py-polars/polars/dataframe/frame.py
Comment (zundertj) [inline py-polars/polars/dataframe/frame.py:1254]: Unless you want a specific error message, you could rely on the lazy loading to raise an error. That would simplify this code to
```python
from polars.dependencies import dataframe_api_compat
return dataframe_api_compat.polars_standard.convert_to_standard_compliant_dataframe(
            self, api_version=api_version
        )
```

Comment (MarcoGorelli) [reply py-polars/polars/dataframe/frame.py]: that's much simpler, thanks @zundertj !

the import is lazy, so it's OK to put it at the top of the file, right?

Comment (zundertj) [reply py-polars/polars/dataframe/frame.py]: Yes it is.

**Pull request merged**
